osu.Game/Online/Chat/DrawableLinkCompiler.cs

```diff
@@ -15,30 +19,109 @@ namespace osu.Game.Online.Chat
     /// <summary>
     /// An invisible drawable that brings multiple <see cref="Drawable"/> pieces together to form a consumable clickable link.
     /// </summary>
-    public class DrawableLinkCompiler : OsuHoverContainer
+    public class DrawableLinkCompiler : OsuHoverContainer, IHasContextMenu
     {
         /// <summary>
         /// Each word part of a chat link (split for word-wrap support).
         /// </summary>
         public List<Drawable> Parts;
 
+        [Resolved(CanBeNull = true)]
+        private OsuGame game { get; set; }
+
+        [Resolved]
+        private BeatmapManager beatmapManager { get; set; }
+
+        private readonly LinkDetails link;
+
         public override bool ReceivePositionalInputAt(Vector2 screenSpacePos) => Parts.Any(d => d.ReceivePositionalInputAt(screenSpacePos));
 
         protected override HoverClickSounds CreateHoverClickSounds(HoverSampleSet sampleSet) => new LinkHoverSounds(sampleSet, Parts);
 
-        public DrawableLinkCompiler(IEnumerable<Drawable> parts)
+        public DrawableLinkCompiler(IEnumerable<Drawable> parts, LinkDetails link)
         {
             Parts = parts.ToList();
+            this.link = link;
         }
 
-        [BackgroundDependencyLoader]
-        private void load(OsuColour colours)
+        [BackgroundDependencyLoader(true)]
+        private void load(OsuColour colours, OsuGame game, BeatmapManager beatmapManager)
         {
             IdleColour = colours.Blue;
+            this.game = game;
+            this.beatmapManager = beatmapManager;
         }
 
         protected override IEnumerable<Drawable> EffectTargets => Parts;
 
+        private BeatmapInfo getBeatmapFromLink(LinkDetails link)
+        {
+            int beatmapId = getBeatmapIdFromLink(link);
+            if (beatmapId != 0)
+                return beatmapManager.QueryBeatmap(b => b.OnlineBeatmapID == beatmapId);
+
+            return null;
+        }
+
+        private BeatmapSetInfo getBeatmapSetFromLink(LinkDetails link)
+        {
+            int setId = getBeatmapSetIdFromLink(link);
+            if (setId != 0)
+                return beatmapManager.QueryBeatmap(b => b.BeatmapSet.OnlineBeatmapSetID == setId)?.BeatmapSet;
+
+            return null;
+        }
+
+        private int getBeatmapIdFromLink(LinkDetails link)
+        {
+            if (int.TryParse(link.Argument.Contains('?') ? link.Argument.Split('?')[0] : link.Argument, out int beatmapId))
```

Comment:
So I'll elaborate on the problem I have with the way this is done. In my view it's unorganised - this is doing link parsing on-the-fly where presumably some places have already *done* such parsing and/or might want to do that parsing on their own again.

What I'd like to see here is introduction of a `BeatmapLink` or a `BeatmapSetLink` model (so *only* and *strictly* data, ideally immutable, too) that is already parsed *once*, at a *higher* level, which has the beatmap ID stored into a sane structure that can just be read here without doing this split & parse locally.